tests/vanilla/effect.test.ts

```diff
@@ -0,0 +1,230 @@
+import { expect, it, vi } from 'vitest'
+import type { Atom, Getter, Setter } from 'jotai/vanilla'
+import { atom, createStore } from 'jotai/vanilla'
+
+type AnyAtom = Atom<unknown>
+type GetterWithPeak = Getter & { peak: Getter }
+type SetterWithRecurse = Setter & { recurse: Setter }
+type Cleanup = () => void
+type Effect = (get: GetterWithPeak, set: SetterWithRecurse) => void | Cleanup
+type Ref = {
+  get: GetterWithPeak
+  set?: SetterWithRecurse
+  cleanup?: Cleanup | null
+  fromCleanup: boolean
+  inProgress: number
+  deps: Set<AnyAtom>
+  init: () => void
+}
+
+function atomSyncEffect(effect: Effect) {
+  const refAtom = atom(
+    () => ({ deps: new Set(), inProgress: 0 }) as Ref,
+    (get) => {
+      const ref = get(refAtom)
+      return () => {
+        ref.cleanup?.()
+        ref.cleanup = null
+        ref.deps.clear()
+      }
+    },
+  )
+  refAtom.onMount = (mount) => mount()
+  const internalAtom = atom((get) => {
+    const ref = get(refAtom)
+    if (!ref.get) {
+      ref.get = ((a) => {
+        ref.deps.add(a)
+        return get(a)
+      }) as Getter & { peak: Getter }
+    }
+    ref.init()
+    ref.deps.forEach(get)
+    if (ref.inProgress > 0) {
+      return
+    }
+    ref.cleanup?.()
+    const cleanup = effectAtom.effect(ref.get!, ref.set!)
+    ref.cleanup =
+      typeof cleanup === 'function'
+        ? () => {
+            try {
+              ref.fromCleanup = true
+              cleanup()
+            } finally {
+              ref.fromCleanup = false
+            }
+          }
+        : null
+  })
+  internalAtom.unstable_onInit = (store) => {
+    const ref = store.get(refAtom)
+    const get = store.get
+    const set = store.set
+    ref.init = () => {
+      if (!ref.get.peak) {
+        ref.get.peak = get
+      }
+      if (!ref.set) {
+        const setter: Setter = (a, ...args) => {
+          try {
+            ++ref.inProgress
+            return set(a, ...args)
+          } finally {
+            --ref.inProgress
+            ref.get(a) // FIXME why do we need this?
+          }
+        }
+        const recurse: Setter = (a, ...args) => {
+          if (ref.fromCleanup) {
+            if (import.meta.env?.MODE !== 'production') {
+              console.warn('cannot recurse inside cleanup')
+            }
+            return undefined as any
+          }
+          return set(a, ...args)
```

Comment:
I think we need to wrap with isRecursing boolean. When not recursing, effect should not infinite loop when setting an watched atom.